Algorithms/Sorters/External/ExternalMergeSorter.cs

```diff
@@ -0,0 +1,127 @@
+using System;
+using System.Collections.Generic;
+
+namespace Algorithms.Sorters.External
+{
+    public class ExternalMergeSorter<T> : IExternalSorter<T>
+    {
+        public void Sort(
+            ISequentialStorage<T> mainMemory,
+            ISequentialStorage<T> temporaryMemory,
+            IComparer<T> comparer)
+        {
+            var originalSource = mainMemory;
+            var source = mainMemory;
+            var temp = temporaryMemory;
+            var totalLength = mainMemory.Length;
+            for (var stripLength = 1L; stripLength < totalLength; stripLength *= 2)
+            {
+                using var left = source.GetReader();
+                using var right = source.GetReader();
+                using var output = temp.GetWriter();
+
+                for (var i = 0L; i < stripLength; i++)
+                {
+                    right.Read();
+                }
+
+                Merge(left, right, output, stripLength, Math.Min(stripLength, totalLength - stripLength), comparer);
+                var step = 2 * stripLength;
+                long rightStripStart;
+                for (rightStripStart = stripLength + step; rightStripStart < mainMemory.Length; rightStripStart += step)
+                {
+                    for (var i = 0L; i < stripLength; i++)
+                    {
+                        left.Read();
+                        right.Read();
+                    }
+
+                    Merge(left, right, output, stripLength, Math.Min(stripLength, totalLength - rightStripStart), comparer);
+                }
+
+                for (var i = 0L; i < totalLength + stripLength - rightStripStart; i++)
+                {
+                    output.Write(right.Read());
+                }
+
+                (source, temp) = (temp, source);
+            }
+
+            if (source == originalSource)
+            {
+                return;
+            }
+
+            using var sorted = source.GetReader();
+            using var dest = originalSource.GetWriter();
+            for (var i = 0; i < totalLength; i++)
+            {
+                dest.Write(sorted.Read());
+            }
+        }
+
+        private static void Merge(
```

Comment:
Codacy found an issue: [Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6254880)